serial: nxp_s32: use instance-based DT macros

manuargue opened the pull request:
At present, many of the NXP S32 shim drivers do not make use of devicetree instance-based macros because the NXP S32 HAL relies on an index-based approach, requiring knowledge of the peripheral instance index during both compilation and runtime, and this index might not align with the devicetree instance index.

The proposed solution in this patch eliminates this limitation by determining the peripheral instance index during compilation through macrobatics and defining the driver's ISR within the shim driver itself.

Note that for some peripheral instances is needed to redefine the HAL macros of the peripheral base address, since the naming is not uniform for all instances.

```
west twister -p s32z270dc2_rtu0_r52@D --device-testing --device-serial=/dev/ttyUSB0 -T tests/drivers/uart/ -T tests/drivers/console/
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.5.0-239-g5a30e6ff013c
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/user/zephyrproject/zephyr/twister-out/testplan.json

Device testing on:

| Platform              | ID   | Serial device   |
|-----------------------|------|-----------------|
| s32z270dc2_rtu0_r52@D |      | /dev/ttyUSB0    |

INFO    - JOBS: 8
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:   30/  30  100%  skipped:   29, failed:    0, error:    0
INFO    - 30 test scenarios (30 test instances) selected, 29 configurations skipped (29 by static filter, 0 at runtime).
INFO    - 1 of 30 test configurations passed (100.00%), 0 failed, 0 errored, 29 skipped with 0 warnings in 20.03 seconds
INFO    - In total 1 test cases were executed, 141 skipped on 1 out of total 627 platforms (0.16%)
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
```

danieldegrasse reviewed (commented):
Very minor nit- can you update commit message to be `drivers: serial: ...`? This better aligns with the commit messages for most updates to `drivers/serial`

manuargue-nxp committed:
serial: nxp_s32: use instance-based DT macros

At present, many of the NXP S32 shim drivers do not make use of
devicetree instance-based macros because the NXP S32 HAL relies on an
index-based approach, requiring knowledge of the peripheral instance
index during both compilation and runtime, and this index might not
align with the devicetree instance index.

The proposed solution in this patch eliminates this limitation by
determining the peripheral instance index during compilation
through macrobatics and defining the driver's ISR within the shim
driver itself.

Note that for some peripheral instances is needed to redefine the
HAL macros of the peripheral base address, since the naming is not
uniform for all instances.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>
